File: java/canard/src/main/java/nl/jessetvogel/canard/parser/Message.java
package nl.jessetvogel.canard.parser;

import nl.jessetvogel.canard.core.Function;

import java.util.List;
import java.util.StringJoiner;

public class Message {

    public enum Status { SUCCESS, FAIL, ERROR }

    public static String create(Status status, String data) {
        return String.format("{\"status\":\"%s\",\"data\":\"%s\"}", status.toString().toLowerCase(), escape(data));
    }

    public static String create(Status status, List<String> data) {
        StringJoiner sj = new StringJoiner(",", "[", "]");
        for(String x : data)
            sj.add("\"" + escape(x) + "\"");
        return String.format("{\"status\":\"%s\",\"data\":%s}", status.toString().toLowerCase(), sj);
    }

    public static String create(Status status, List<String> keys, List<List<String>> data) {
        int n = keys.size();
        StringJoiner sjList = new StringJoiner(",", "[", "]");
        for(List<String> result : data) {
            StringJoiner sjMap = new StringJoiner(",","{","}");
            for(int i = 0;i < n; ++i)
                sjMap.add("\"" + escape(keys.get(i)) + "\":\"" + escape(result.get(i)) + "\"");
            sjList.add(sjMap.toString());
        }
        return String.format("{\"status\":\"%s\",\"data\":%s}", status.toString().toLowerCase(), sjList);
    }

    private static String escape(String raw) {
        StringBuilder sb = new StringBuilder();
        for (char ch : raw.toCharArray()) {
            if (ch == '\"') sb.append("\\\"");
            else if (ch == '\\') sb.append("\\\\");
            else if (ch == '\b') sb.append("\\b");
            else if (ch == '\f') sb.append("\\f");
            else if (ch == '\n') sb.append("\\n");
            else if (ch == '\r') sb.append("\\r");
            else if (ch == '\t') sb.append("\\t");
            else if (ch >= 0x20 && ch <= 0x7E) sb.append(ch);
            else sb.append(String.format("\\u%04X", (int) ch));
        }
        return sb.toString();
    }

}

File: cpp/canard/parser/message.cpp
//
// Created by Jesse Vogel on 01/07/2021.
//

#include <sstream>
#include <iomanip>
#include "message.h"

std::string Message::escape(const std::string &raw) {
    std::ostringstream ss;
    for (char c : raw) {
        if (c == '\"') ss << "\\\"";
        else if (c == '\\') ss << "\\\\";
        else if (c == '\b') ss << "\\b";
        else if (c == '\f') ss << "\\f";
        else if (c == '\n') ss << "\\n";
        else if (c == '\r') ss << "\\r";
        else if (c == '\t') ss << "\\t";
        else if (c >= 0x20 && c <= 0x7E) ss << c;
        else ss << "\\u" << std::hex << std::setw(4) << std::setfill('0') << (int) c;
    }
    return ss.str();
}

const std::string status_strings[] = {"success", "fail", "error"};

std::string Message::create(MessageStatus status, const std::string &data) {
    std::ostringstream ss;
    ss << "{\"status\":\"" << status_strings[status] << "\",\"data\":\"" << escape(data) << "\"}";
    return ss.str();
}

std::string Message::create(MessageStatus status, const std::vector<std::string> &data) {
    std::ostringstream ss;
    ss << '[';
    bool first = true;
    for (const std::string &s : data) {
        if (!first) ss << ',';
        first = false;
        ss << '"' << escape(s) << '"';
    }
    ss << ']';
    return create(status, ss.str());
}

std::string
Message::create(MessageStatus status, const std::vector<std::string> &keys, const std::vector<std::string> &values) {
    std::ostringstream ss;
    size_t n = keys.size();
    ss << '{';
    bool first = true;
    for (int i = 0; i < n; ++i) {
        if (!first) ss << ',';
        first = false;
        ss << '"' << escape(keys[i]) << "\":\"" << escape(values[i]) << '"';
    }
    ss << '}';
    return create(status, ss.str());
}
